ngraph/core/reference/include/ngraph/runtime/reference/interpolate.hpp

```diff
@@ -0,0 +1,592 @@
+                ///
+                /// \param input_data pointer to input data
+                /// \param out pointer to memory block for output data
+                void linear_onnx_func(const T* input_data, T* out);
+
+                /// \brief Calculates cubic interpolation.
+                ///
+                /// \param input_data pointer to input data
+                /// \param out pointer to memory block for output data
+                void cubic_func(const T* input_data, T* out);
+
+                /// \brief Calculates 'nearest' mode of interpolation.
+                ///
+                /// \param input_data pointer to input data
+                /// \param out pointer to memory block for output data
+                void nearest_func(const T* input_data, T* out);
+            };
+
+            template <typename T>
+            void InterpolateEval<T>::linear_func(const T* input_data, T* out)
+            {
+                size_t input_rank = m_input_data_shape.size();
+                size_t num_of_axes = m_axes.size();
+
+                auto info = helper.get_info_for_linear_mode();
+
+                CoordinateTransform output_transform(m_out_shape);
+                CoordinateTransform input_transform(m_input_data_shape);
+
+                for (const Coordinate& output_coord : output_transform)
+                {
+                    auto icoords_data = helper.get_icoords(output_coord);
+
+                    float summa = 0.0f;
+                    float wsum = 0.0f;
+
+                    CoordinateTransform indices{info.shape_for_indeces};
+                    for (const auto& index : indices)
+                    {
+                        auto inner_result =
+                            helper.inner_calculation(output_coord, icoords_data, info, index);
+                        if (!inner_result.condition)
+                        {
+                            continue;
+                        }
+
+                        wsum += inner_result.w;
+                        summa += inner_result.w *
+                                 static_cast<float>(
+                                     input_data[input_transform.index(inner_result.inner_coord)]);
+                    }
+
+                    if (wsum == 0.0f)
+                    {
+                        out[output_transform.index(output_coord)] = T{};
+                    }
+                    else
+                    {
+                        out[output_transform.index(output_coord)] = static_cast<T>(summa / wsum);
+                    }
+                }
+            }
+
+            template <typename T>
+            void InterpolateEval<T>::linear_onnx_func(const T* input_data, T* out)
+            {
+                size_t input_rank = m_input_data_shape.size();
+                size_t num_of_axes = m_axes.size();
+
+                assert((input_rank == 2) || (input_rank == 4));
```

Comment:
@lazarevevgeny These asserts are expected? I didn't find any limitations for 2d and 4d shapes in the spec.

Reply:
This is an ONNX operation limitation. I would remove this assert in our code.

Reply:
These limitations is in lines 366-367 of the file `Interpolate_4.md`. If we delete this limitations, and ONNX `Resize` in some future opset will be extended using other way, we will need to create new operation.